libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -2057,15 +2057,22 @@ string YulUtilFunctions::copyValueArrayStorageToStorageFunction(ArrayType const&
 			solAssert(!_fromType.isValueType(), "");
 		templ("functionName", functionName);
 		templ("resizeArray", resizeArrayFunction(_toType));
-		templ("arrayLength",arrayLengthFunction(_fromType));
+		templ("arrayLength", arrayLengthFunction(_fromType));
 		templ("panic", panicFunction(PanicCode::ResourceError));
 		templ("srcDataLocation", arrayDataAreaFunction(_fromType));
 		templ("dstDataLocation", arrayDataAreaFunction(_toType));
 		templ("srcStride", to_string(_fromType.storageStride()));
 		unsigned itemsPerSlot = 32 / _toType.storageStride();
 		templ("itemsPerSlot", to_string(itemsPerSlot));
 		templ("multipleItemsPerSlotDst", itemsPerSlot > 1);
-		bool sameType = _fromType.baseType() == _toType.baseType();
+		bool sameType = *_fromType.baseType() == *_toType.baseType();
+		if (auto functionType = dynamic_cast<FunctionType const*>(_fromType.baseType()))
+		{
+			solAssert(functionType->equalExcludingStateMutability(
+				dynamic_cast<FunctionType const&>(*_toType.baseType())
+			));
```

Comment:
Can't this be made stricter, i.e., we should assert that `toType`'s state mutability is greater than or equal to that of `fromType`? So a `view` function can be assigned to a function without state mutability, but not the other way around.

Reply:
I don't think this falls into the responsibility of the code generator. If we at some time allow to have such a type conversion (for example through explicit conversion), the code generator should not fail. What we care about here is that the data representation is the same.

Reply:
Yeah, the only situation where the data representation can be different is if the function kind is different (external vs internal for example) and the `equalExcludingStateMutability()` assert already takes care of that so I think the current asserts are enough.